watcher.go

```diff
@@ -0,0 +1,67 @@
+package consul
+
+import (
+	"context"
+	"strconv"
+)
+
+type WatcherFunc func([]KeyData)
+
+type Watcher struct {
+	Client *Client
+}
+
+var (
+	DefaultWatcher = &Watcher{Client: DefaultClient}
+)
+
+func Watch(ctx context.Context, key string, handler WatcherFunc) {
+	DefaultWatcher.Watch(ctx, key, handler)
+}
+
+func (w *Watcher) Watch(ctx context.Context, key string, handler WatcherFunc) {
+	w.watching(ctx, key, handler, nil)
+}
+
+func WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	DefaultWatcher.WatchPrefix(ctx, prefix, handler)
+}
+
+func (w *Watcher) WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	q := Query{Param{
+		Name:  "recurse",
+		Value: "",
+	}}
+	w.watching(ctx, prefix, handler, q)
+}
+
+func updateParam(q Query, new Param) Query {
+	ret := make(Query, len(q))
+	copy(ret, q)
+	for i, p := range q {
+		if new.Name == p.Name {
+			ret[i] = new
+			return ret
+		}
+	}
+	return append(ret, new)
+}
+
+func (w *Watcher) watching(ctx context.Context, key string, handler WatcherFunc, q Query) {
+	q = updateParam(q, Param{Name: "index", Value: "0"})
+	resp := []KeyData{}
+	path := "/v1/kv/" + key
+	hdr, err := w.Client.do(ctx, "GET", path, q, nil, &resp)
+	if err != nil {
+		// What to do?
+	}
```

Comment:
It seems like the only difference between this first path and the loop is `index=0`, could we rely on the loop only to avoid duplicating code?

Reply:
I didn't want to send an erroneous event on the handler before anything's changed but thinking more about it, this implementation requires the caller to handle changes so they need the initial anyway.

Updating.